### PR TITLE
Uplift `{Closure,Coroutine,CoroutineClosure}Args` and friends to `rustc_type_ir`

### PR DESCRIPTION
**Pull request by compiler-errors**
Part of converting the new solver's `structural_traits.rs` to be interner-agnostic.

I decided against aliasing `ClosureArgs<TyCtxt<'tcx>>` to `ClosureArgs<'tcx>` because it seemed so rare. I could do so if desired, though.

r? lcnr

### CODE REVIEW

File: compiler/rustc_type_ir/src/ty_kind/closure.rs [hunk 1]
```diff
@@ -0,0 +1,689 @@
...
+                tupled_inputs_ty
+                    .tuple_fields()
+                    .into_iter()
+                    .chain(closure_tupled_upvars_ty.tuple_fields()),
+            ),
+        }
+    }
+}
+
+struct FoldEscapingRegions<I: Interner> {
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/ty_kind/closure.rs:514]: existing: i feel like this folder should have a comment

**Review by lcnr — COMMENTED** [review]
r=me with or without nit

**Comment by compiler-errors** [comment]
Sorry for the CI failures. Tried to be smart with an assertion, but it was misguided.

@bors r=lcnr

**Comment by compiler-errors** [comment]
intra doc links :(

@bors r-

**Pull request merged**